### DISCUSSION TITLE
Improvements switching from Coral USB to Mini PCIe (didn't boot with 4 cards!)

### DISCUSSION THREAD

**Discussion by natm** [blakeblackshear/frigate#597] [General]
I have a reasonable size Frigate instance at home:

- 13 cameras
- 2 feeds per camera:
- Main feed 1920x1080 @ 25 FPS: clips, record, RTMP
- Sub feed 704x576 @ 6 FPS: detect

Since 2019 I've been using a Coral USB, but after finally selling my old Hikvision NVR on eBay over Christmas, thought I'd invest in reducing my inference speed. It has sat around 130ms with 13 cameras (when I run a single camera via USB it is about 10ms).

Yesterday I removed the USB coral and added a [Mini PCIe card](https://coral.ai/products/pcie-accelerator) in a [wifi adapter](https://uk.banggood.com/Mini-WiFi-Wan-802_11n-PCI-e-to-PCI-e-Wireless-Expansion-Card-Adapter-Convertor-p-1013592.html) carrier.

![Screenshot from 2021-01-24 11-48-41](https://user-images.githubusercontent.com/1790618/105629305-23670400-5e3a-11eb-80a3-7e07fd29fddd.png)

My inference speed dropped from 130ms to 6.7ms, total detection FPS no longer seems throttled, before it would max out at around 8 FPS. During snow last night on the PCIe card it got to around 90 FPS.

**Last 24 hours,** you can see when the USB was swapped to PCIe...

![Screenshot from 2021-01-24 11-37-15](https://user-images.githubusercontent.com/1790618/105629275-ef8bde80-5e39-11eb-8b35-e4e27c226061.png)

**Last 6 hours**

![Screenshot from 2021-01-24 11-44-39](https://user-images.githubusercontent.com/1790618/105629291-016d8180-5e3a-11eb-93eb-915a776ae9f4.png)

I currently have 1 installed in my home Supermicro Xeon server, when I tried adding multiple cards it wouldn't boot, so I need to investigate the PCIe configuration in the BIOS further.

![Screenshot from 2021-01-24 11-50-23](https://user-images.githubusercontent.com/1790618/105629365-7b9e0600-5e3a-11eb-8129-acb7865e171a.png)

With 4 PCIe cards! didn't boot :-(

![Screenshot from 2021-01-24 11-51-49](https://user-images.githubusercontent.com/1790618/105629387-9a040180-5e3a-11eb-962c-161139c54b02.png)

**Comment by jmedlin** [comment]
Very interesting.  I just realized that my ASUS Chromebox CN60 that I'm running this on has a Mini PCIe slot on it and ordered a coral to shove in there.

**Comment by paularmstrong** [comment]
Why did you need to add it onto a wifi adapter carrier?

**Reply by natm** [reply]
The Coral Mini PCIe card would not fit directly into the "full size" PCIe slots on the motherboard.

**Comment by apaperclip** [comment]
very cool and interesting!  
How you are you distributing work among the multiple edge tpus?  I think the recommended detect FPS is 5, that may offer some more capacity.

**Comment by ehn** [comment]
@natm Were you able to get the multiple cards working?

**Reply by natm** [reply]
No, the server I was trying them in was hosting other containers which I didn't want to take offline for too long.

In the end I only managed to get one PCIe coral to boot, could be my BIOS config though.

**Reply by ozett** [reply]
seems that no great benefit of multiple CORALs and you can stick to one CORAL
https://github.com/blakeblackshear/frigate/issues/2281#issuecomment-974038686

**Comment by Yonny24** [comment]
Nice! Just today I've been looking into alternatives as there are no USB corals.

So if I have a mini pcie this will have excellent performance and easy to set up with Frigate?
The M.2 could also be an option if I have a spare slot?

I read that the temp needs to be controlled carefully which I wasn't aware of. How do you manage this?

https://coral.ai/products/pcie-accelerator/

**Reply by ozett** [reply]
i read the CORAL temps with that linux command line
and visualize with munin. may this help get temps from the coral?
https://github.com/blakeblackshear/frigate/issues/2281#issuecomment-974152438

**Comment by grimmaldus** [comment]
Have you experienced any issues with these cheap Chinese adapters? For some reason, they cause my server to shut down unexpectedly. When I remove them, everything works fine. I'm trying to find a replacement, but it's been quite difficult to find one within the EU. Any recommendations would be greatly appreciated!
